Feat(eos_designs): Generate CVP Tags

chetryan opened the pull request:
## Change Summary

This change adds a new feature: to configure Tags on CloudVision Portal and associate them with Devices and Interfaces.
This will allow for:
- Correct and amendable generation of the Topology in CVP
- Make Filtering in Dashboards easier
- Allow for the specification of custom CVP Tags within the data model.

## Related Issue(s)

No issues were raised for this new feature.

## Component(s) name

`arista.avd.eos_designs`
`arista.avd.cloudvision`

## Proposed changes
<!--- Describe your changes in detail -->
<!--- Describe data model implemented for new features -->
- Add `cloudvision_tags` data structure to the generated `eos_designs` structured configs
- Add new `arista.avd.cloudvision` role to deploy these tags to cloudvision.

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

- Run `make webdoc-up` to host a local copy of the documentation that can be read on a browser.

- Run the `molecule test -s cloudvision_tags` scenario. This will generate the structured config with cv_tags.

**To run tests on a cloudvision instance**
- Create a lab with 1 cloudvision and the same devices as defined in cloudvision_tags scenario
- Ensure devices are onboarded to cloudvision.
- Set the follow environment variables:
  - CVP_ADDRESS
   - CVP_USERNAME
   -  CVP_PASSWORD
- Run the `molecule test -s cloudvision_tags` scenario. This will generate the structured config with cv_tags **and apply the tags to cloudvision.**

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- Provide feedback on documentation of new feature.
- Provide feedback on naming of new role and data model.
- Test on existing projects and topologies.

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

ClausHolbechArista committed:
Getting started

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into cvp_tags

chetryan committed:
Feat(cvp_tags): Add topology and peer tags

chetryan committed:
Feat(cvp_tags): Update github action

chetryan committed:
Feat(cvp_tags): update actions

chetryan committed:
Fix(eos_designs): Restore removed line.

chetryan committed:
Feat(cvp_tags): Add interface description

chetryan committed:
Fix(cvp_tags): Fix molecule create play

chetryan committed:
Feat(cvp_tags): Add default hint type and custom

chetryan committed:
Fix(cvp_tags): Update Schema

chetryan committed:
Fix(cvp_tags): Extend list of invalid custom tags

chetryan committed:
Feat(cvp_tags): Add push to CVP

chetryan committed:
Feat(cvp_tags): Add documentation generation

chetryan committed:
Feat(cvp_tags): Support is_deployed and int tags

chetryan committed:
Fix(cvp_tags): skip CVP in molecule test

When running the molecule test without the Env variables
for CVP, skip trying to push the tags to cvp.

chetryan committed:
Fix(cvp_tags): Fix the generate_release

chetryan committed:
Docs(cvp_tags): Add Documentation for cvp_tags

chetryan committed:
WIP: move generation to eos_designs

chetryan committed:
WIP: Added push to cloudvision

chetryan committed:
WIP

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into cvp_tags

chetryan committed:
Fix: Regenerate after merge

chetryan committed:
Docs: Update documentation and refactor some vars

chetryan committed:
Docs: Add Documentation

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan commented:
After discussing internally, the data model will be changed to use the `name` key instead of `label` .

chetryan committed:
Fix(CI): Update the CI for the new feature

chetryan committed:
Fix(eos_designs): Rename 'label' to 'name'

chetryan committed:
Fix(cloudvision): Remove unused action module

chetryan committed:
Merge branch 'devel' into cvp_tags

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+    auto_create: true
+  delegate_to: "{{ cloudvision_node }}"
+  when:
+    - not hostvars[inventory_hostname].is_deployed is arista.avd.defined(false)
```
```suggestion
    - is_deployed is not arista.avd.defined(false)
```
same elsewhere

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+    filename: "{{ structured_dir }}/{{ inventory_hostname }}.{{ avd_structured_config_file_format }}"
+
+- name: "Generate CloudVision Tag Data"
+  set_fact:
```
instead of setting this fact, you can just define the same data directly on the argument of the cv_tag_v3 below.

chetryan replied:
Agreed. I had this in place with a debug msg that I have now removed. Good catch thank you

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/cloudvision_tags.schema.yml:
```diff
@@ -0,0 +1,41 @@
+type: dict
+keys:
+  cloudvision_tags:
+    required: false
```
All the required:false are not needed. That is the default.

chetryan replied:
Understood - updated the schema.

chetryan committed:
Fix(eos_designs): Re run molecule tests

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/input-variables.md:
````diff
@@ -593,6 +593,117 @@ roles/eos_designs/docs/tables/system-settings.md
+??? example "Example: Overriding default topology type"
+
+    ```yaml
+      # Some example here
````
Instead of showing the example inside the inventory structure, I would prefer to just show it as regular host/group vars, since that is the more likely place to use this.
so just
```yaml
cloudvision_tags_topology_type: edge
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/input-variables.md:
```diff
@@ -593,6 +593,117 @@ roles/eos_designs/docs/tables/system-settings.md
 roles/eos_designs/docs/tables/cloudvision-settings.md
 --8<--
 
+## CloudVision Tags
```
Please move this to a how-to document at keep this section brief with just the table of supported variables.

chetryan replied:
Agreed, done.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/input-variables.md:
```diff
@@ -593,6 +593,117 @@ roles/eos_designs/docs/tables/system-settings.md
+    - **Note 2: If the key specified in `field` is not found, the tag is not generated. This avoids having a lot of empty tags.**
+
+**Generating Tags for Devices**
+--8<--
```
Add a blank line between the heading and table

chetryan replied:
This document has be refactored to point to the How-To docs and only include a simple table.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/tables/cloudvision-tags-device-custom.md:
````diff
@@ -0,0 +1,15 @@
+=== "YAML"
+
+    ```yaml
+    cloudvision_tags_device_custom:
````
Maybe call it `cloudvision_custom_device_tags`?

chetryan replied:
I chose to prefix all the keys with `cloudvsision_tags_` so that it's easier to identify them then if the `_tag` part was at the end. I agree it's not easy to read though.

chetryan replied:
All variables have now been changed to use the prefix `cv_tags_` as discussed in the chat.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/tables/cloudvision-tags-device-generate.md:
````diff
@@ -0,0 +1,15 @@
+=== "YAML"
+
+    ```yaml
+    cloudvision_tags_device_generate:
````
Maybe call it `cloudvision_generate_device_tags`

chetryan replied:
This has been renamed to `cv_tags_generate_device` to keep the prefix discussed/agreed on in the chat.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/tables/cloudvision-tags-topology-type.md:
````diff
@@ -0,0 +1,11 @@
+=== "YAML"
+
+    ```yaml
+    cloudvision_tags_topology_type: <str>
````
maybe `cloudvision_topology_tags_device_type` ? since there can come other types.

chetryan replied:
All variables have now been changed to use the prefix `cv_tags_` as discussed in the chat. In this case the variable is now called `cv_tags_topology_type`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/tables/cloudvision-tags-device-generate.md:
````diff
@@ -0,0 +1,15 @@
+    ```yaml
+    cloudvision_tags_device_generate:
+      - name: <str>
+        field: <str>
````
Here we support the a dot-notated path to the data right? Maybe `data_path` is a better key? At least improve the description to make this clear, maybe with a few examples.

chetryan replied:
Agreed and updated the schema.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/input-variables.md:
```diff
@@ -593,6 +593,117 @@ roles/eos_designs/docs/tables/system-settings.md
 
+## CloudVision Tags
+
+`arista.avd.eos_designs` can generate CloudVision Tags that can be applied to interfaces and/or devices. These tags can be used during Topology generateion, or used in searches/filters to select devices based on tags values.
```
I would like to have a general key to enable/disable this entire feature, so existing deployments would not get the tags in their structured config. From the 500+ files changes, I can see that it affects everything right now :)

chetryan replied:
Added the `avd_generate_cloudvision_tags` key that defaults to False. If it is set to True, the tags will be generated. Re-ran the molecule scenarios as well to confirm this behaviour.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+
+    def __init__(self, hostvars: dict, shared_utils: SharedUtils):
+        # Init a display object in order to later issue warnings
+        self.ansible_display = Display()
```
please avoid this. Also clean up the commented debug lines below.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+        hints = [self._topology_hint_dc, self._topology_hint_fabric, self._topology_hint_pod, self._topology_hint_type, self._topology_hint_rack]
+        for hint in hints:
+            if hint:
+                device_tags.append(hint)
```
```suggestion
device_tags = [hint for hint in hints if hint]
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+        """
+        Generate the data structure `cloudvision_tags`.
+        """
+        device_tags = []
```
```suggestion
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+            if hint:
+                device_tags.append(hint)
+
+        for custom_tag in get(self.shared_utils.hostvars, "cloudvision_tags_device_custom", []):
```
We should add `convert_to_lower_case: true` to the schema since tags are not always case sensitive on CV (varies unfortunately, so better to be safe)

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+                raise AristaAvdError(f"{generate_tag['name']} is Invalid. System Tags cannot be overriden")
+            if type(value) in [list, dict]:
+                raise AristaAvdError(f"The field {generate_tag['field']} appears to be a {type(value).__name__}. This is not supported for cloudvision fields.")
+            if value:
```
This means a value of `false` is not included. Please try to avoid these "loose" boolean checks around the code, unless you know the type is a bool. Otherwise use something like `is not None`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+                pass
+
+        interface_tags = []
+        for link in get(self.shared_utils.hostvars, "ethernet_interfaces", []):
```
You should use "your own" hostvars placed in self._hostvars

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+    @cached_property
+    def _topology_hint_type(self) -> dict:
+        """
+        Retrun the topology hint type for the device.
```
```suggestion
        Return the topology hint type for the device.
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+        return result
+
+    @staticmethod
+    def tag_dict(name, value):
```
```suggestion
    def tag_dict(name, value) -> dict:
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+            if type(value) in [list, dict]:
+                raise AristaAvdError(f"The field {generate_tag['field']} appears to be a {type(value).__name__}. This is not supported for cloudvision fields.")
+            if value:
+                tags.append({"name": generate_tag["name"], "value": str(value)})
```
use self.tag_dict?

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+        for link in get(self.shared_utils.hostvars, "ethernet_interfaces", []):
+            generated_interface_tags = self._interface_tags(link)
+            if generated_interface_tags:
+                interface_tags.append(generated_interface_tags)
```
If you change the `_interface_tags()` to always return a list, you can simplify the code to
```suggestion
            interface_tags.extend(self._interface_tags(link))
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+            if type(value) in [list, dict]:
+                raise AristaAvdError(f"The field {generate_tag['field']} appears to be a {type(value).__name__}. This is not supported for cloudvision fields.")
+            if value:
+                device_tags.append({"name": generate_tag["name"], "value": str(value)})
```
use self.tags_dict?

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+            if custom_tag["name"].lower() not in INVALID_CUSTOM_DEVICE_TAGS:
+                device_tags.append(custom_tag)
+            else:
+                raise AristaAvdError(f"{custom_tag['name']} is Invalid. System Tags cannot be overriden")
```
Please provide more context, since the use does not know which part of the code emits this error. Same for other errors. Maybe just "CloudVision tag ..."

ClausHolbechArista commented on ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py:
```diff
@@ -154,6 +154,14 @@ def evpn_prevent_readvertise_to_server(self: SharedUtils) -> bool:
     def dc_name(self: SharedUtils) -> str | None:
         return get(self.hostvars, "dc_name")
 
+    @cached_property
+    def fabric_name(self: SharedUtils) -> str | None:
+        return get(self.hostvars, "fabric_name")
```
We will not run if fabric_name is not set, so you can add `required=True` to the get and then remove the | None

ClausHolbechArista commented on ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/node_type_keys.py:
```diff
@@ -42,16 +38,10 @@
-            "type": "overlay-controller",
-            "default_evpn_role": "server",
-        },
+        {"key": "super_spine", "type": "super-spine", "cloudvision_tags_topology_type": "core"},
```
please add comma to the end of the last value so `black` will unfold these

chetryan replied:
Added the ',' so these lines are unfolded.

ClausHolbechArista commented on ansible_collections/arista/avd/requirements-dev.txt:
```diff
@@ -10,7 +10,7 @@ pre-commit>=2.9.2
 identify>=1.4.20
 docker
-molecule>=3.2.0
+molecule>=5.0.1
```
This is a major change, we have to discuss

chetryan replied:
Removed this change. I was having some issues with particular molecule versions. We can discuss, but does not need to go into this PR.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/tables/cloudvision-tags-topology-type.md:
```diff
@@ -0,0 +1,11 @@
+=== "Table"
```
Please add `documentation_options` with a table name to your schema fragments, so you can gather them in a single table where it makes sense.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,178 @@
+        """
+        Return the topology fabric hint tag.
+        """
+        if not self.shared_utils.fabric_name:
+            # self.ansible_display.warning(msg=f"'fabric_name' not found for {self.shared_utils.hostname}")
+            return None
```
If you update so fabric_name is required, this can be simplified.

ClausHolbechArista reviewed (changes requested):
Some comments on code and naming. All naming suggestions are only suggestions so only follow them if you agree :)

chetryan committed:
WIP: Make feature off by default

chetryan committed:
WIP: Rerun molecule tests

chetryan committed:
Fix: Revert changes to mpls example

chetryan committed:
WIP: Update documentation.

chetryan committed:
Fix: Update eos_cli_config_gen schema

chetryan committed:
Docs & Refactor: Refactor avdstructured config

Also update some markdown documentation

chetryan committed:
Merge branch 'devel' into cvp_tags

ClausHolbechArista commented on ansible_collections/arista/avd/README.md:
```diff
@@ -32,6 +32,7 @@ This repository provides content for Arista's **arista.avd** collection. The fol
 - [**arista.avd.eos_config_deploy_cvp**](roles/eos_config_deploy_cvp/README.md) - Deploys intended configuration via CloudVision.
 - [**arista.avd.eos_config_deploy_eapi**](roles/eos_config_deploy_eapi/README.md) - Deploys intended configuration via eAPI.
 - [**arista.avd.cvp_configlet_upload**](roles/cvp_configlet_upload/README.md) - Uploads configlets from a local folder to CloudVision Server.
+- [**arista.avd.cloudvision**](roles/cloudvision/README.md) - to Cloudvision, .
```
Looks like something is missing here.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/inventory/inventory.yml:
```diff
@@ -0,0 +1,83 @@
+          ansible_python_interpreter: $(which python3)
+          configlet_dir_var: 'configlets'
+          configlets_prefix_var: 'AVD'
+          configlet_file_extension: 'conf'
```
I don't think we need any of these.
```suggestion
```

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/inventory/inventory.yml:
```diff
@@ -0,0 +1,83 @@
+    CVP:
+      hosts:
+        cv_server:
+          ansible_httpapi_host: "{{ lookup('ansible.builtin.env', 'CVP_ADDRESS') }}"
```
```suggestion
```

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/inventory/inventory.yml:
```diff
@@ -0,0 +1,83 @@
+            DC1_SUPER_SPINES:
+              hosts:
+                dc1-super-spine1:
+                  ansible_host: 172.16.1.211
```
Nothing in this scenario will need to communicate with the devices directly, so please remove all these ansible_host entries.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/inventory/inventory.yml:
```diff
@@ -0,0 +1,83 @@
+                      ansible_host: 172.16.1.151
+                    dc1-pod1-leaf2c:
+                      ansible_host: 172.16.1.152
+                      cloudvision_tags_topology_type: edge
```
Please use group_vars or host_vars instead of setting things in the inventory.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/inventory/inventory.yml:
```diff
@@ -0,0 +1,83 @@
+---
```
In general try to limit the size of this inventory and vars to the minimum required to test all functionality. It will run quicker and there will be less redundant maintenance.

chetryan replied:
I have attempted to reduce the inventory to the minimum topology that could be tested on cloudvision and still have 5-clos design for L3LS. Kindly let me know if it can be simplified further.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/molecule.yml:
```diff
@@ -0,0 +1,50 @@
+    pre_build_image: true
+    managed: false
+    groups:
+      - FABRIC
```
```suggestion
platforms:
  - name: dummy
    managed: false
```

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/cloudvision_tags/molecule.yml:
```diff
@@ -0,0 +1,50 @@
+  env:
+    # Custom pythonpath to allow loading custom ip_addressing and descriptions modules.
+    # See CUSTOM_PYTHON_MODULES_TESTS group vars for details.
+    PYTHONPATH: ${PYTHONPATH}:${MOLECULE_SCENARIO_DIRECTORY}/custom_modules
```
```suggestion
```

ClausHolbechArista commented on ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/misc.py:
```diff
@@ -154,6 +154,14 @@ def evpn_prevent_readvertise_to_server(self: SharedUtils) -> bool:
     def dc_name(self: SharedUtils) -> str | None:
         return get(self.hostvars, "dc_name")
 
+    @cached_property
+    def fabric_name(self: SharedUtils) -> str:
+        return get(self.hostvars, "fabric_name", required=True)
+
+    @cached_property
+    def rack(self: SharedUtils) -> str | None:
+        return get(self.switch_data_combined, "rack")
+
```
Please update other references to these variables to pick up from here.

chetryan replied:
The only times I have found these referenced were in the new `AvdStructuredConfigTags` class in `avdstructuredconfig.py` . References to 'fabric_name' in eos_designs_facts.py seem to refer to the `fabric_name` variable passed to the eos_designs_role.

chetryan replied:
Changed all references to these variables, which turned out to be in snmp.

ClausHolbechArista commented on ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_shared_utils/node_type_keys.py:
```diff
@@ -13,12 +13,7 @@
-            "default_evpn_role": "server",
-            "default_ptp_priority1": 20,
-        },
+        {"key": "spine", "type": "spine", "default_evpn_role": "server", "default_ptp_priority1": 20, "cloudvision_tags_topology_type": "spine"},
```
Add comma at the end

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/README.md:
````diff
@@ -0,0 +1,129 @@
+
+Here is an overview with the key configured in the YAML inventory:
+
+```yaml
````
I think we can make a more clear example here without ansible_port etc.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/README.md:
```diff
@@ -0,0 +1,129 @@
+
+### Tasks
+
+1. Load the `structurd_config` and parse the CloudVision tags.
```
```suggestion
1. Load the `structured_config` created by `eos_designs` and parse the CloudVision tags.
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/README.md:
```diff
@@ -0,0 +1,129 @@
+### Tasks
+
+1. Load the `structurd_config` and parse the CloudVision tags.
+2. Generate the CloudVision tags data structure.
```
Is this step still there?

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/defaults/main/main.yml:
I think you can remove the schema defaults here, but you should add `structured_dir` since you reference that in the tasks

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/tasks/main.yml:
```diff
@@ -0,0 +1,24 @@
+---
+- name: "Load tag information from strucutred config"
```
```suggestion
- name: "Load tag information from structured config"
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/cloudvision/tasks/main.yml:
```diff
@@ -0,0 +1,24 @@
+      'device': "{{ inventory_hostname }}",
+      'device_tags': "{{ cloudvision_tags.device_tags }}",
+      'interface_tags': "{{ cloudvision_tags.interface_tags }}"
+    } ]
```
It will be more readable if you just use YAML formatting here.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/defaults/main/main.yml:
```diff
@@ -25,3 +25,6 @@ avd_structured_config_file_format: "yml"
 avd_data_validation_mode: "warning"
+
+# Enable/Disable CloudVision tag generation
+avd_generate_cloudvision_tags: False
```
Please just implement the default in your code and the documentation and avoid using the Ansible defaults files.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/docs/role-configuration.md:
```diff
@@ -22,3 +22,15 @@ roles/eos_designs/docs/tables/role-documentation-output-settings.md
+This can be set in the host/group vars to influence all or some of the devices. Alternatively it can be set in the playbook.
+
+!!! note
+        **Refer to the [how-to](./how-to/cloudvision-tags.md) documentation for more details on how to use this feature.**
```
```suggestion
    **Refer to the [how-to](./how-to/cloudvision-tags.md) documentation for more details on how to use this feature.**
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,179 @@
+
+    """
+
+    def __init__(self, hostvars: dict, shared_utils: SharedUtils):
+        super().__init__(hostvars=hostvars, shared_utils=shared_utils)
```
```suggestion
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py:
```diff
@@ -0,0 +1,179 @@
+from ansible_collections.arista.avd.plugins.plugin_utils.avdfacts import AvdFacts
+from ansible_collections.arista.avd.plugins.plugin_utils.errors import AristaAvdError
+from ansible_collections.arista.avd.plugins.plugin_utils.utils import get
+from ansible_collections.arista.avd.roles.eos_designs.python_modules.underlay.utils import UtilsMixin
```
Please avoid importing from other modules. Move to shared utils if you still need something from this one.

chetryan commented on .github/workflows/pull-request-management.yml:
```diff
@@ -302,6 +306,49 @@ jobs:
+        avd_scenario:
+          - 'cloudvision_tags'
+        ansible_version:
+          - 'ansible-core>=2.12.6,<2.15.0,!=2.13.0'
```
```suggestion
          - 'ansible-core>=2.12.6,<2.15.0,!=2.13.0'
          - 'ansible-core==2.12.6'
```

chetryan commented on .github/workflows/pull-request-management.yml:
```diff
@@ -302,6 +306,49 @@ jobs:
+          - 'cloudvision_tags'
+        ansible_version:
+          - 'ansible-core>=2.12.6,<2.15.0,!=2.13.0'
+        # Also test minimum ansible version for one scenario.
+        include:
+          - avd_scenario: 'cloudvision_tags'
+            ansible_version: 'ansible-core==2.12.6'
```
```suggestion
```

chetryan committed:
Update .github/workflows/pull-request-management.yml

chetryan committed:
Update .github/workflows/pull-request-management.yml

chetryan committed:
Update ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
Update ansible_collections/arista/avd/roles/eos_designs/docs/role-configuration.md

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
Update ansible_collections/arista/avd/roles/cloudvision/tasks/main.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
WIP: feedback from PR

chetryan committed:
Merge branch 'cvp_tags' of https://github.com/chetryan/ansible-avd into cvp_tags

chetryan committed:
WIP: PR feedback

chetryan committed:
WIP: PR feedback

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
Merge branch 'cvp_tags' of https://github.com/chetryan/ansible-avd into cvp_tags

chetryan committed:
Fix: Feedback from PR

Turn on CloudVision tags on a single device in eos_designs_unit_tests
Turn off CloudVision tags for a single node in cloudvision_tags
Change arista.avd.cloudvision to work with the delegate option
Update documentation.
Move the variable definitions from the inventory to group/host_vars.

chetryan committed:
Fix: Remove commented out lines

carlbuchmann committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into cvp_tags

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
DOCS(cloudvision): Update docs

chetryan committed:
FEAT(cloudvision): rename the fields to improve readability

chetryan committed:
FIX: use failed_when instead of ignore_errors

chetryan committed:
FIX: eos_designs_unit_tests

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
FEAT: rename cloudvision_ prefic to cv_

chetryan committed:
FEAT: continue rename of cloudvision_ prefix

chetryan committed:
Fix: Remove faulty delegation test

Could not find a test to ensure the role had been delegated
that works 100% of the time, so it's better to remove it than
give false sense of security to user.

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into cvp_tags

chetryan committed:
FIX(cloudvision): pipeline error

Fix pipeline erro for the molecule test for cloudvision

chetryan committed:
FIX: make device_tags or interfac_tags optional

chetryan committed:
Fix: Update molecule test to cover more scenarios

chetryan committed:
FIX: Update the molecule

chetryan committed:
FIX(eos_designs): Rename field to data_path

chetryan committed:
FIX(eos_designs): Use the shared_utils.rack information

chetryan committed:
FIX(eos_designs) Simplify molecule test topology

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan committed:
Fix(eos_designs): fix molecule test

carlbuchmann commented:
Requires the following CVP PR to be merged and new ansible-cvp released: https://github.com/aristanetworks/ansible-cvp/pull/679

chetryan committed:
Merge branch 'devel' into cvp_tags

chetryan commented:
> Requires the following CVP PR to be merged and new ansible-cvp released: [aristanetworks/ansible-cvp#679](https://github.com/aristanetworks/ansible-cvp/pull/679)

That CVP PR has now been merged. Is it time to revisit this PR?

ClausHolbechArista commented:
Closing this, but keeping the branch, so we can revive as needed later. I will branch from this to implement the subset generating the tags. The tags will then be pushed as part of the new CV integration in #3466
